### PR TITLE
implemented event inheritance tree

### PR DESCRIPTION
**Pull request by kouks**
**References issue**
#73 

**Describe the pull request**
Added a simple script that triggers more abstract listeners. Solves the problem referenced in the issue.

**Additional context**
I included some questions about the `Optional<T>` so that I don't forget about it. Please do not merge with those.

### CODE REVIEW

File: lib/core/src/structures/optional/Optional.ts [hunk 1]
```diff
@@ -49,6 +49,7 @@ export interface Optional<T> {
    * @param {any} e
    * @return {T}
    */
+  // ASK: Should this be Error for type scrictness?
```
Comment (bausano) [inline lib/core/src/structures/optional/Optional.ts]: Thought about it, kind of liked that you're not limited to exceptions only.

File: lib/core/src/structures/optional/Optional.ts [hunk 2]
```diff
@@ -58,6 +59,7 @@ export interface Optional<T> {
    * @param {any} def
    * @return {any}
    */
+  // ASK: Should this be T for type scrictness?
```
Comment (bausano) [inline lib/core/src/structures/optional/Optional.ts]: Yes should be

File: lib/core/src/structures/optional/Optional.ts [hunk 3]
```diff
@@ -67,6 +69,7 @@ export interface Optional<T> {
...
```
Comment (bausano) [inline lib/core/src/structures/optional/Optional.ts]: Yes should be

File: lib/core/src/structures/optional/Optional.ts [hunk 4]
```diff
@@ -77,6 +80,7 @@ export interface Optional<T> {
    * @param {none?: () => any} none
    * @return {any}
    */
+  // ASK: Should this be T for type scrictness?
```
Comment (bausano) [inline lib/core/src/structures/optional/Optional.ts]: No, should be any.

**Comment by kouks** [comment]
Removed comments and edited the `Optional.unwrapOr` and `Optional.unwrapOrElse` signatures.

**Pull request merged**